ui/src/app/machines/views/MachineDetails/MachineStorage/AvailableStorageTable/AddPartition/AddPartition.tsx

```diff
@@ -0,0 +1,137 @@
+  diskId: Disk["id"];
+  systemId: Machine["system_id"];
+};
+
+const AddPartitionSchema = Yup.object().shape({
+  filesystemType: Yup.string(),
+  mountOptions: Yup.string(),
+  mountPoint: Yup.string().when("filesystemType", {
+    is: (val) => !!val,
+    then: Yup.string()
+      .matches(/^\//, "Mount point must start with /")
+      .required("Mount point is required if filesystem type is defined"),
+  }),
+  partitionSize: Yup.number()
+    .required("Size is required")
+    .when("unit", {
+      is: "2",
+      then: Yup.number().min(5, "Partition must be at least 5MB"),
+      otherwise: Yup.number().min(0, "Size must greater than 0"),
+    }),
+  unit: Yup.string().required(),
+});
+
+export const AddPartition = ({
+  closeExpanded,
+  diskId,
+  systemId,
+}: Props): JSX.Element | null => {
+  const dispatch = useDispatch();
+  const machine = useSelector((state: RootState) =>
+    machineSelectors.getById(state, systemId)
+  );
+  const { creatingPartition } = useSelector((state: RootState) =>
+    machineSelectors.getStatuses(state, systemId)
+  );
+  const previousCreatingPartition = usePrevious(creatingPartition);
+  const saved = !creatingPartition && previousCreatingPartition;
+
+  // Close the form when partition has successfully been created.
+  // TODO: Check for machine-specific error, in which case keep form open.
+  // https://github.com/canonical-web-and-design/maas-ui/issues/1842
+  useEffect(() => {
+    if (saved) {
+      closeExpanded();
+    }
+  }, [closeExpanded, saved]);
+
+  if (machine && "disks" in machine && "supported_filesystems" in machine) {
+    const filesystemOptions = machine.supported_filesystems.map(
+      (filesystem) => ({
+        label: filesystem.ui,
+        value: filesystem.key,
+      })
+    );
+    const disk = machine.disks.find((disk) => disk.id === diskId);
+    const partitionName = disk
+      ? `${disk.name}-part${(disk.partitions?.length || 0) + 1}`
+      : "partition";
+
+    return (
+      <FormikForm
+        buttons={FormCardButtons}
+        cleanup={machineActions.cleanup}
+        initialValues={{
+          filesystemType: "",
+          mountOptions: "",
+          mountPoint: "",
+          partitionSize: "",
+          unit: "3",
+        }}
+        onCancel={closeExpanded}
+        onSaveAnalytics={{
+          action: "Add partition",
+          category: "Machine storage",
+          label: "Add partition",
+        }}
+        onSubmit={(values: AddPartitionValues) => {
+          const {
+            filesystemType,
+            mountOptions,
+            mountPoint,
+            partitionSize,
+            unit,
+          } = values;
+          const size = partitionSize * Math.pow(1000, Number(unit));
+          const params = {
+            blockId: diskId,
+            filesystemType: filesystemType || undefined,
+            mountOptions: (filesystemType && mountOptions) || undefined,
+            mountPoint: (filesystemType && mountPoint) || undefined,
```

Comment:
Instead of setting this to `undefined` you could use some `if` conditions below to add the values or spread the values onto the object e.g. `...(filesystemType && mountPoint && {mountPoint})`

Reply:
Went with the spread method, thanks!